packages/whenable/src/whenable.js

```diff
@@ -0,0 +1,111 @@
+// @ts-check
+import { makePromiseKit } from '@endo/promise-kit';
+import { M } from '@endo/patterns';
+import { makeTagged } from '@endo/pass-style';
+
+/**
+ * @param {import('@agoric/base-zone').Zone} zone
+ */
+export const prepareWhenableKits = zone => {
+  /** WeakMap<object, any> */
+  const whenableV0ToEphemeral = new WeakMap();
+
+  /**
+   * Get the current incarnation's promise kit associated with a whenableV0.
+   *
+   * @param {import('./types.js').WhenablePayload['whenableV0']} whenableV0
+   * @returns {import('@endo/promise-kit').PromiseKit<any>}
+   */
+  const findCurrentKit = whenableV0 => {
+    let pk = whenableV0ToEphemeral.get(whenableV0);
+    if (pk) {
+      return pk;
+    }
+
+    pk = makePromiseKit();
+    pk.promise.catch(() => {}); // silence unhandled rejection
+    whenableV0ToEphemeral.set(whenableV0, pk);
+    return pk;
+  };
+
+  /**
+   * @param {'resolve' | 'reject'} kind
+   * @param {import('./types.js').WhenablePayload['whenableV0']} whenableV0
+   * @param {unknown} value
+   */
+  const settle = (kind, whenableV0, value) => {
+    const kit = findCurrentKit(whenableV0);
+    const cb = kit[kind];
+    if (!cb) {
+      return;
+    }
+    whenableV0ToEphemeral.set(whenableV0, harden({ promise: kit.promise }));
+    cb(value);
+  };
+
+  const makeWhenableInternalsKit = zone.exoClassKit(
+    'WhenableInternalsKit',
+    {
+      whenableV0: M.interface('WhenableV0', {
+        shorten: M.call().returns(M.promise()),
+      }),
+      settler: M.interface('Settler', {
+        resolve: M.call().optional(M.any()).returns(),
+        reject: M.call().optional(M.any()).returns(),
+      }),
+    },
+    () => ({}),
+    {
+      whenableV0: {
+        /**
+         * @returns {Promise<any>}
+         */
+        shorten() {
+          return findCurrentKit(this.facets.whenableV0).promise;
+        },
+      },
+      settler: {
+        /**
+         * @param {any} [value]
+         */
+        resolve(value) {
+          const { whenableV0 } = this.facets;
+          settle('resolve', whenableV0, value);
```

Comment:
I think this would be more amenable to static analysis if `settle` were replaced with a function that returned the promise kit.
```js
  const getPromiseResolvers = whenableV0 => {
    const kit = findCurrentKit(whenableV0);
    if (kit.resolve) {
      // Resolution is a one-time event, so forget the resolve/reject functions.
      whenableV0ToEphemeral.set(whenableV0, harden({ promise: kit.promise }));
    }
    return kit;
  };
```
```suggestion
          const { resolve } = getPromiseResolvers(whenableV0);
          if (resolve) resolve(value);
```

Reply:
This is a completely encapsulated change, yes?